Detector/DetFCChhTrackerTkLayout/src/TkLayoutEndcap_Geo.cpp

```diff
@@ -47,119 +43,122 @@ static DD4hep::Geometry::Ref_t createTkLayoutTrackerEndcap(DD4hep::Geometry::LCD
     Component xDisc = static_cast<Component>(xDiscColl);
-    currentZ = xDisc.z() - dimensions.zmin() - 0.5 * (dimensions.zmax() - dimensions.zmin());
-    PlacedVolume placedDiscVolume = envelopeVolume.placeVolume(discVolume, DD4hep::Geometry::Position(0, 0, currentZ));
-    placedDiscVolume.addPhysVolID("disc", discCounter);
-    ++discCounter;
+    currentZ = xDisc.z() - dimensions.zmin() - envelopeThickness;
+    DetElement disc_det(worldDetElement, "disc" + std::to_string(discCounter), discCounter);
+    // generate rings and place in  discs
+    for (DD4hep::XML::Collection_t xRingColl(xFirstDiscRings, _U(ring)); nullptr != xRingColl; ++xRingColl) {
+      Component xRing = static_cast<Component>(xRingColl);
+      Component xRingModules = xRing.child("modules");
+      Component xModuleOdd = xRingModules.child("moduleOdd");
+      Component xModuleEven = xRingModules.child("moduleEven");
+      Component xModuleProperties = xRing.child("moduleProperties");
+      Component xModulePropertiesComp = xModuleProperties.child("components");
+      Component xSensorProperties = xRing.child("sensorProperties");
+      Volume moduleVolume("module",
+                          DD4hep::Geometry::Trapezoid(0.5 * xModuleProperties.attr<double>("modWidthMin"),
+                                                      0.5 * xModuleProperties.attr<double>("modWidthMax"),
+                                                      0.5 * xModuleProperties.attr<double>("modThickness"),
+                                                      0.5 * xModuleProperties.attr<double>("modThickness"),
+                                                      0.5 * xSensorProperties.attr<double>("sensorLength")),
+                          lcdd.material("Air"));
+      // place components in module
+      double integratedCompThickness = 0;
+      int componentCounter = 0;
+      for (DD4hep::XML::Collection_t xCompColl(xModulePropertiesComp, _U(component)); nullptr != xCompColl; ++xCompColl) {
+        Component xComp = static_cast<Component>(xCompColl);
+        Volume componentVolume("component",
+                               DD4hep::Geometry::Trapezoid(0.5 * xModuleProperties.attr<double>("modWidthMin"),
+                                                           0.5 * xModuleProperties.attr<double>("modWidthMax"),
+                                                           0.5 * xComp.thickness(),
+                                                           0.5 * xComp.thickness(),
+                                                           0.5 * xSensorProperties.attr<double>("sensorLength")),
+                               lcdd.material(xComp.materialStr()));
+        PlacedVolume placedComponentVolume = moduleVolume.placeVolume(
+            componentVolume,
+            DD4hep::Geometry::Position(
+                0, integratedCompThickness - 0.5 * xModuleProperties.attr<double>("modThickness"), 0));
```

Comment:
Hi Andre,

it is on the list! I am still trying to get the fitter working in this PR, should have written a comment to explain that it was still WIP. But thanks!